src/lib/route.js

```diff
@@ -187,20 +242,67 @@ function dataToRoute (data, currentEpoch) {
     additionalInfo: data.additional_info,
     targetPrefix: data.target_prefix,
     addedDuringEpoch: currentEpoch
-  })
+  }, data.paths)
 }
 
 /**
- * @param {Array} listA
- * @param {Array} listB
- * @returns {Integer} the number of items that listA and listB share
+ * @param {Route} routeA
+ * @param {Route} routeB
+ * @returns {Boolean} whether routeA and routeB can join without forming a loop
  */
-function intersect (listA, listB) {
-  let common = 0
-  for (const itemA of listA) {
-    if (listB.indexOf(itemA) !== -1) common++
+function canJoin (routeA, routeB) {
+  const visited = {}
+  // These routes would be concatenated as:
+  //  routeA.sourceLedger, routeA.nextLedger, [[ routeA.paths ]], routeA.destinationLedger
+  //                                                              === routeB.sourceLedger, routeB.nextLedger, [[ routeB.paths ]], routeB.destinationLedger
+  //
+  //  routeC.sourceLedger, routeC.nextLedger, [[ routeA.paths ** [routeA.destinationLedger, routeB.nextLedger] ** routeB.paths ]], routeC.destinationLedger
+  // (the ** tries to express that any path from routeA can be combined with any path from routeB)
+
+  // These three should always be different from each other:
+  const fixedLedgers = [routeA.sourceLedger, routeA.nextLedger, routeB.destinationLedger]
+  if (routeA.destinationLedger !== routeB.sourceLedger) {
+    // routes are not adjacent, can't join them
+    return false
+  }
+  // If routeA has third ledger, add it:
+  if (routeA.destinationLedger !== routeA.nextLedger) {
+    fixedLedgers.push(routeA.destinationLedger)
   }
-  return common
+  // If routeB has third ledger, add it:
+  if (routeB.destinationLedger !== routeB.nextLedger) {
+    fixedLedgers.push(routeB.nextLedger)
+  }
+
+  // Now we have 3, 4, or 5 fixed ledgers; check that they all differ:
+  for (let ledger of fixedLedgers) {
+    if (visited[ledger]) return false
+    visited[ledger] = true
+  }
+
+  // Check for intersections between routeA's paths and visited:
+  // Remember paths are the alternative options between routeA.nextLedger and routeA.destinationLedger.
+  //
+  for (let path of routeA.paths) {
+    for (let ledger of path) {
+      if (visited[ledger]) return false
+    }
+  }
+
+  // Now add all ledgers from routeA's paths to visited:
```

Comment:
Because it's allowed to have a path ['a.', 'b.'] as well as a path ['a.', 'c.']; the double occurrence of 'a.' is in alternative paths, so it's not a loop as long as neither 'a.' nor 'b.' nor 'c.' appear in the list of fixed ledgers, or in the paths of routeB.